test/helpers/utils.go

```diff
@@ -122,15 +123,35 @@ func WithTimeout(body func() bool, msg string, config *TimeoutConfig) error {
 	done := time.After(config.Timeout * time.Second)
 	ticker := time.NewTicker(config.Ticker * time.Second)
 	defer ticker.Stop()
-	if body() {
-		return nil
+	status := make(chan bool)
+
+	var mutex = &sync.Mutex{}
+	// If the function take more than the config.Ticker the function will be
+	// not executed.
+	bodyLock := false
+	setBodyLock := func(val bool) {
+		mutex.Lock()
+		bodyLock = val
+		mutex.Unlock()
 	}
+	WithLock := func(body func() bool) {
+		if bodyLock == true {
+			return
+		}
+		setBodyLock(true)
+		defer func() { setBodyLock(false) }()
+		status <- body()
+	}
+
+	go WithLock(body)
 	for {
 		select {
-		case <-ticker.C:
-			if body() {
+		case res := <-status:
+			if res {
 				return nil
 			}
+		case <-ticker.C:
+			go WithLock(body)
 		case <-done:
 			return fmt.Errorf("Timeout reached: %s", msg)
```

Comment:
There's no guarantees the go routines scheduled are all completed / canceled when the timeout is reached, this might influence the following tests.

Reply:
No, there is no way, but if a timeout happens the AfterEach, JustAferEach, AfterFailed, AfterAll takes cares of keep the state as it was before. So there is no problem with that.

Reply:
Can you guarantee that in the future we will never use the error of WithTimeout for other than failing the test? How does the After functions keep the state as it was before? For example, the `body` function can perform any operation. What if the `body` is doing a `kubectl apply`?
1 - the timeout is set for 10 seconds
2 - the go routine blocks for 2 minutes for some reason
3 - the timeout returns an error with `Timeout reached`
4 - the next test is running and in the middle of the test, the `kubectl apply` of the previous `body` function will run. This can create undesirable behavior. Why can't we change the WithTimeout with the `context`? And all functions using `WithTimeout` should be using this new context.

This doesn't need to be done in this PR, we can create a new `WithTimeout` function that accepts a `context` called `WithTimeoutCtx` and move all usages of the `WithTimeout` to `WithTimeoutCtx` slowly over time.

Reply:
> Can you guarantee that in the future we will never use the error of WithTimeout for other than failing the test? How does the After functions keep the state as it was before? For example, the body function can perform any operation. What if the body is doing a kubectl apply?

I mean, if the function return true the `withTimeout` body function finish. If
the test fails, the AfterEach functions will keep all as the desired state, so
test result does not matter at all, is the afterEach who cares about the cilium
state.


>1 - the timeout is set for 10 seconds
>2 - the go routine blocks for 2 minutes for some reason
>3 - the timeout returns an error with Timeout reached
>4 - the next test is running and in the middle of the test, the kubectl apply of the previous body function will run. This can create undesirable behavior. Why can't we change the WithTimeout with the context? And all functions using WithTimeout should be using this new context.

Maybe I'm wrong, but I do not think that any `WithTimeout` body made a kubectl
apply or something similar, all WithTimeout function only does validation that
things are in a desired state (Pods ready, waiting until policy is applied,
etc..) if not the logic of the `WithTimeout` will not make sense to me at all.

Yes, we can have a `WithTimeoutCtx` but I do not see a real value at all, if we
need to run a background command that apply something can be run with `kub.ExecContext` or
`kub.ExecPodContext` that are designed for that.

I understand your points, but with the current test, and the structure that the
framework follows I do not see the winning point at all.

Reply:
@aanm ^^